ICRC-31: extract icrc31_list_identities (#103)
dfinity/wg-identity-authentication: topics/icrc_31_list_identities.md

# ICRC-31: List Identities (ICRC-25 Extension)

[![Status Badge](https://img.shields.io/badge/STATUS-DRAFT-ffcc00.svg)](https://github.com/orgs/dfinity/projects/31)

<!-- TOC -->

- [ICRC-31: List Idenities](#icrc-31-list-identities)
    - [Summary](#summary)
    - [Method](#method)
    - [Request](#request-2)
    - [Response](#response-2)

>     - [Errors](#errors-2)

Reviewer: This toc seems to be out of sync with the headings here.

    - [Message Processing](#message-processing-2)
    - [Example](#example-2)

## Summary

The purpose of the `icrc31_list_identities` message is for the relying party to receive information about the identities managed by the signer.

## Method

**Name and Scope:** `icrc31_list_identities`

> **Prerequisite:** Active session with granted permission for method name/scope.

Reviewer: ```suggestion
**Prerequisite:** Active session with granted permission scope `icrc31_list_identities` or `*`.
```

Reviewer: Nit: I would make this as clear as possible to the reader what the exact permission needs to be.

### Request

**`version` (`text`):** The version of the standard used. If the signer does not support the version of the request, it must send the `"VERSION_NOT_SUPPORTED"` error in response.

**Example RPC Request**:

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "icrc31_list_identities",
  "params": {
    "version": "1"
  }
}
```

### Response

`version` (`text`): The version of the standard used. It must match the `version` from the request.

`identities`: A list of identities the user has selected to share with the relying party.

- `publicKey` (`blob`): The DER-encoded public key associated with the identity, derived in accordance with one of [the signature algorithms supported by the IC](https://internetcomputer.org/docs/current/references/ic-interface-spec/#signatures). The public key can be used to [derive a self-authenticating principal](https://internetcomputer.org/docs/current/references/ic-interface-spec/#principal).

**Example RPC Response:**

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "version": "1",
    "identities": [
      {

>         "publicKey": "MFYwEAYHKoZIzj0CAQYFK4EEAAoDQgAEOTdHYwpFTr/oPXOfLQcteymk8AQE41VwPQ1W7Xpm0Zt1AY4+5aOnMAbAIjXEchxPuGbPWqPqwntXMPs3w4rOaA=="

Reviewer: Here I would opt for transferring principals as the method is about _identities_ rather than public keys. Also, I think in most cases the relying party is only interested in the principal.
So my suggestion: `publicKey` -> `principal`

Also you might want to consider to just have an array of principals (textual representation so the type would be `string[]`) for most simplicity unless you expect the returned objects to be extended in the future. But allowing extensions on the returned identities also goes against the spirit of this restructuring that aims to extend the standards by introducing more method (rather than altering existing ones).

Author: Good point; I'm looking through my cross-chain glasses. That's why I've decided to move on with an object per identity. But I agree with you. Maybe we should be even more explicit and call this method `icrc_31_get_principal_ids?

Reviewer: > Maybe we should be even more explicit and call this method `icrc_31_get_principal_ids?

I would suggest just `icrc_(list|get)_principals` which corresponds to the term that the [interface spec](https://internetcomputer.org/docs/current/references/ic-interface-spec/#principal) is using.
